File: Tests/AlternatorTests.cs
using System;
using System.Text.Patterns;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Tests {
	[TestClass]
	public class AlternatorTests {
		[TestMethod]
		public void Constructor() {
			Pattern _ = (Pattern)"Hello" | "Goodbye";
		}

		[TestMethod]
		public void Consume() {
			Pattern Alternator = (Pattern)"Hello" | "Goodbye";
			Result Result;

			Result = Alternator.Consume("Hello");
			ResultAssert.Captures("Hello", Result);

			Result = Alternator.Consume("Goodbye");
			ResultAssert.Captures("Goodbye", Result);

			Result = Alternator.Consume("!");
			ResultAssert.Fails(Result);

			ResultAssert.Fails(Alternator.Consume("How are you?"));

			Pattern ChainAlternator = (Pattern)"Hello" | "Hi" | "Howdy";

			Result = ChainAlternator.Consume("Hello");
			ResultAssert.Captures("Hello", Result);

			Result = ChainAlternator.Consume("Hi");
			ResultAssert.Captures("Hi", Result);

			Result = ChainAlternator.Consume("Howdy");
			ResultAssert.Captures("Howdy", Result);
		}

		[TestMethod]
		public void Equal() {
			Pattern Alternator = (Pattern)"Hello" | "Goodbye";
			Assert.IsTrue(Alternator.Equals("Hello"));
			Assert.IsTrue(Alternator.Equals("Goodbye"));
			Assert.IsFalse(Alternator.Equals("World"));
		}
	}
}

File: Tests/Patterns/AlternatorTests.fs
namespace Tests.Patterns

open System
open Stringier.Patterns
open Microsoft.VisualStudio.TestTools.UnitTesting

[<TestClass>]
type AlternatorTests() =

    [<TestMethod>]
    member _.``constructor`` () =
        let _pattern = "Hello" || "Goodbye"
        ()

    [<TestMethod>]
    member _.``consume`` () =
        let pattern = "Hello" || "Goodbye"
        let mutable result = Result()

        result <- pattern.Consume("Hello")
        ResultAssert.Captures("Hello", result)

        result <- pattern.Consume("Goodbye")
        ResultAssert.Captures("Goodbye", result)

        result <- pattern.Consume("!")
        ResultAssert.Fails(result)

        result <- pattern.Consume("How are you?")
        ResultAssert.Fails(result)

        let chainPattern = "Hello" || "Hi" || "Howdy"

        result <- chainPattern.Consume("Hello")
        ResultAssert.Captures("Hello", result)

        result <- chainPattern.Consume("Hi")
        ResultAssert.Captures("Hi", result)

        result <- chainPattern.Consume("Howdy")
        ResultAssert.Captures("Howdy", result)

        result <- chainPattern.Consume("Goodbye")
        ResultAssert.Fails(result)

    [<TestMethod>]
    member _.``boolean-or still works`` () =
        Assert.IsTrue(true || true)
        Assert.IsTrue(false || true)
        Assert.IsTrue(true || false)
        Assert.IsFalse(false || false)

    [<TestMethod>]
    member _.``neglect`` () =
        let pattern = not ("Hello" || "Goodbye")
        ResultAssert.Fails(pattern.Consume("Hello"))
        ResultAssert.Fails(pattern.Consume("Goodbye"))
        ResultAssert.Captures("World", pattern.Consume("Worldeater"))